### PR TITLE
HTTP: Refactoring nxt_h1p_peer_header_send()

### PR DESCRIPTION
**Pull request by hongzhidao**
Previously, proxy request was constructed based on the `r->target` field. However, r->target will remain unchanged in the future, even in cases of URL rewriting.
To accommodate this, I refactored the handling of the proxy target.

### CODE REVIEW

**Review by ac000 — COMMENTED** [review]
Hi Zhidao,

> Previously, proxy request was constructed based on the `r->target`
field. However, r->target will remain unchanged in the future,
even in cases of URL rewriting.
To accommodate this, I refactored the handling of the proxy target.

This doesn't really tell me _why_ `r->target` remaining unchanged will be an issue.

File: src/nxt_h1proto.c [hunk 1]
```diff
@@ -90,6 +90,8 @@ static void nxt_h1p_peer_connect(nxt_task_t *task, nxt_http_peer_t *peer);
 static void nxt_h1p_peer_connected(nxt_task_t *task, void *obj, void *data);
 static void nxt_h1p_peer_refused(nxt_task_t *task, void *obj, void *data);
 static void nxt_h1p_peer_header_send(nxt_task_t *task, nxt_http_peer_t *peer);
+static nxt_int_t nxt_h1p_peer_request_target(nxt_http_request_t *r,
+    nxt_str_t *target);
```
Comment (ac000) [inline src/nxt_h1proto.c:94]: This change seems unrelated to the rest. I.e I don't see you adding this function for example.

Does adding this get rid of some warning introduced by this commit?

File: src/nxt_h1proto.c [hunk 2]
```diff
@@ -654,6 +656,8 @@ nxt_h1p_header_process(nxt_task_t *task, nxt_h1proto_t *h1p,
     r->target.start = h1p->parser.target_start;
     r->target.length = h1p->parser.target_end - h1p->parser.target_start;
 
+    r->quoted_target = h1p->parser.quoted_target;
+
```
Comment (ac000) [inline src/nxt_h1proto.c:660]: What does `quoted_target` mean?

**Comment by hongzhidao** [comment]
Hi Andrew,
For now, `r->target` is changed during URL rewriting. It will be constant because we need to keep `$request_uri` unchanged. Here's the change.
https://github.com/nginx/unit/pull/1162/files#diff-8f22b5e7ac8aa80082722fd822e96917284058f7aa449aeffec406a5e0520beeL80 
Let me know if it's necessary to add something like above to the commit log.

File: src/nxt_h1proto.c [hunk 3]
```diff
@@ -2353,6 +2362,61 @@ nxt_h1p_peer_header_send(nxt_task_t *task, nxt_http_peer_t *peer)
...
+    size = r->path->length + encode * 2 + 1 + r->args->length;
+
+    target->start = nxt_mp_nget(r->mem_pool, size);
+    if (target->start == NULL) {
+        return NXT_ERROR;
+    }
+
+    if (r->quoted_target) {
+        p = (u_char *) nxt_encode_complex_uri(target->start, r->path->start,
+                                              r->path->length);
```
Comment (ac000) [inline src/nxt_h1proto.c:2406]: This begged the question, will a `size_t` always be big enough to hold a `uintptr_t`.

While I don't believe the C standard makes any guarantees that that will be the case, I think in practice it will be, at least for any system we care about.

On 64bit Linux (and macOS) they are defined like

```
$ echo | gcc -E -xc -include 'stddef.h' - | grep size_t
typedef long unsigned int size_t;
$ echo | gcc -E -xc -include 'stdint.h' - | grep uintptr_t
typedef unsigned long int uintptr_t;
```

On 32bit Linux

```
$ echo | gcc -E -xc -include 'stddef.h' - | grep size_t
typedef unsigned int size_t;
$ echo | gcc -E -xc -include 'stdint.h' - | grep uintptr_t
typedef unsigned int uintptr_t;
```

File: src/nxt_http_parse.h [hunk 4]
```diff
@@ -61,12 +61,10 @@ struct nxt_http_request_parse_s {
 
     /* target with "/." */
     uint8_t                   complex_target;         /* 1 bit */
-#if 0
     /* target with "%" */
     uint8_t                   quoted_target;          /* 1 bit */
     /* target with " " */
     uint8_t                   space_in_target;        /* 1 bit */
-#endif
```
Comment (ac000) [inline src/nxt_http_parse.h:69]: Curious why we are only now enabling these two things?

**Comment by ac000** [comment]
> For now, `r->target` is changed during URL rewriting. It will be
> constant because we need to keep `$request_uri` unchanged. Here's the
> change.

Yes, but I'm more asking _why_ it remaining unchanged will be an issue
for the proxy going forward.

**Comment by hongzhidao** [comment]
> This doesn't really tell me why r->target remaining unchanged will be an issue.
Yes, but I'm more asking why it remaining unchanged will be an issue
for the proxy going forward.

There is no issue currently for proxy.

Ideally, the `$request_uri` will always correspond to `r->target` which happens in `nxt_http_variables.c`.
During URL rewriting, `r->target` is changed, and the same is for `$request_uri`.
We decided to make `$request_uri` constant, which means `r->target` will also be unchanged, even in case of URL rewriting. The logic of touching `r->target` will be removed.
https://github.com/nginx/unit/pull/1162/files#diff-8f22b5e7ac8aa80082722fd822e96917284058f7aa449aeffec406a5e0520beeL80

To make it short, the `r->target` should be designed to be constant, but Unit needs to pass a changeable URL to upstream server in proxy module. So proxy can't depend on `r->target`.

About the quoted_target, take the following as an example.
```
client request: /blah%25blah?a=b
r->path: /blah%blah
r->target:/blah%2Fblah?a=b
```
`quoted_target` means it contains percent-encoding char like `%25`.
If it's quoted, it means we need to encode `r->path`.
And we also need to take into arguments account, if there is argument, we need to append it.
The logic is something like this:
```
if (quoted_target || args->length > 0) {
    we need to generate a new URL.
}
```

File: src/nxt_http_parse.h (hunk 4 above)
Comment (hongzhidao) [reply src/nxt_http_parse.h:69]: See my new comment, but it looks like `space_in_target` is unused. Will check.

File: src/nxt_h1proto.c (hunk 2 above)
Comment (hongzhidao) [reply src/nxt_h1proto.c:660]: See my new comment.
It happens during http parsing.

File: src/nxt_h1proto.c (hunk 1 above)
Comment (hongzhidao) [reply src/nxt_h1proto.c:94]: Sorry, I don't understand what you mean.
`nxt_h1p_peer_request_target()` is called in `nxt_h1p_peer_header_send()`.

Comment (ac000) [reply src/nxt_h1proto.c:94]: Ah, I see it.

<rant>
If ordered our functions correctly we wouldn't need all these prototypes...
</rant>

**Comment by ac000** [comment]
> To make it short, the `r->target` should be designed to be constant,
> but Unit needs to pass a changeable URL to upstream server in proxy
> module. So proxy can't depend on `r->target`.

Right this is what I was after.

So, the proxy needs to see the re-written URL... gotcha.

That's a key bit of information missing from the commit message, please
add it.

**Comment by ac000** [comment]
> About the quoted_target, take the following as an example.

> ```
> client request: /blah%25blah?a=b
> r->path: /blah%blah
> r->target:/blah%2Fblah?a=b
> ```

> quoted_target means it contains percent-encoding char like `%25`.  If
> it's quoted, it means we need to encode `r->path`.  And we also need
> to take into arguments account, if there is argument, we need to
> append it.

OK, so _quoted_target_ means _percent_encoded_. It's just badly named.

**Comment by hongzhidao** [comment]
I think it's just a refactoring patch, there are no functional changes introduced.
When proxy was introduced, there is no rewrite feature at that time, it makes sense to use `r->target` which is equal to `r->path` + `r->args`. So what I did here is to use a more proper way of creating proxy sending header.

Btw, I'm thinking about separating quoted_target related code into a new commit.

File: src/nxt_h1proto.c (hunk 3 above)
Comment (hongzhidao) [reply src/nxt_h1proto.c:2406]: I replaced it with `nxt_uint_t` that referred to `nxt_http_return.c`.
I feel both `nxt_uint_t` and `size_t` are correct since the max size is limited by the `large_header_buffer_size` option that is validated by the Unit JSON. And the max number length is 14.
```
#define NXT_CONF_MAX_NUMBER_LEN    14
```

File: src/nxt_h1proto.c (hunk 1 above)
Comment (hongzhidao) [reply src/nxt_h1proto.c:94]: Hmm, the other way is to put the function definition before the call, both the two ways are fine.

Comment (ac000) [reply src/nxt_h1proto.c:94]: > Hmm, the other way is to put the function definition before the call, both the two ways are fine.

If you mean this, then yes...

```c
static void func_c(void)
{
        /* ... */
}

static void func_b(void)
{
        func_c();
}

static void func_a(void)
{
        func_b();
}

int main(void)
{
        func_a();

        return 0;
}
```

Following that pattern you rarely need to provide prototypes as the compiler will see the function definition _before_ it's used.

**Review by ac000 — COMMENTED** [review]
**EDIT:** This is in relation to [HTTP: Introduce quoted target marker in HTTP parsing](https://github.com/nginx/unit/pull/1214/commits/28ed5912fcd2f55eb7dd07e4b77e34a4000fa6e4)

Perhaps a bit of text in the commit message about _why_ we need this now and we didn't need it before...

File: src/nxt_h1proto.c (hunk 3 above)
Comment (ac000) [reply src/nxt_h1proto.c:2406]: That makes me more nervous! What if that constant ever increases... I think it was **safer** with `size_t`.

**Comment by ac000** [comment]
> I think it's just a refactoring patch, there are no functional changes
> introduced.

But it's making use of the `quoted_target` thing which it wasn't before.
Does that not equate to a change in behaviour?

File: src/nxt_h1proto.c (hunk 3 above)
Comment (hongzhidao) [reply src/nxt_h1proto.c:2406]: I kept `size_t`.

**Comment by hongzhidao** [comment]
> **EDIT:** This is in relation to [HTTP: Introduce quoted target marker in HTTP parsing](https://github.com/nginx/unit/pull/1214/commits/28ed5912fcd2f55eb7dd07e4b77e34a4000fa6e4)
> 
> Perhaps a bit of text in the commit message about _why_ we need this now and we didn't need it before...

I don't know exactly why, because this part of the code was commented out, probably because it wasn't needed at the time.
But I added a note about when it can be used, that's its intention.

**Comment by hongzhidao** [comment]
> **EDIT:** This is in relation to [HTTP: Introduce quoted target marker in HTTP parsing](https://github.com/nginx/unit/pull/1214/commits/28ed5912fcd2f55eb7dd07e4b77e34a4000fa6e4)
> 
> Perhaps a bit of text in the commit message about _why_ we need this now and we didn't need it before...

I don't know exactly why, because this part of the code was commented out, probably because it wasn't needed at the time.
But I added a note about when it can be used, that's its intention.

**Comment by hongzhidao** [comment]
> > I think it's just a refactoring patch, there are no functional changes
> > introduced.
> 
> But it's making use of the `quoted_target` thing which it wasn't before. Does that not equate to a change in behaviour?

I split it into two commits, for the first one, maybe it's a code change rather than refactoring.
And I think it's a refactoring for the second commit.

**Comment by ac000** [comment]
For me the second patch ((HTTP: Refactoring nxt_h1p_peer_header_send()](https://github.com/nginx/unit/pull/1214/commits/95727a1bae2224b1aa1a2c245583a1429eb14342)) is not a refactoring.

I would either split it up to do the refactor bit, i.e without the `quoted_target` stuff, then a second patch making use of the `quoted_target` stuff.

We didn't use `quoted_target` before, a _refactor_ should not suddenly start using it I think...

Or just admit it isn't a refactor and adjust the commit message...

**Comment by hongzhidao** [comment]
> Or just admit it isn't a refactor and adjust the commit message...

Ok, the same for the first one. Removed "No functional changes".

File: src/nxt_h1proto.c [hunk 5]
```diff
@@ -2353,6 +2362,61 @@ nxt_h1p_peer_header_send(nxt_task_t *task, nxt_http_peer_t *peer)
...
+
+    if (r->quoted_target) {
+        encode = nxt_encode_complex_uri(NULL, r->path->start,
+                                        r->path->length);
+    } else {
+        encode = 0;
+    }
+
+    size = r->path->length + encode * 2 + 1 + r->args->length;
+
```
Comment (ac000) [inline src/nxt_h1proto.c:2398]: Just want to confirm that the required calculation is (taking operator precedence into account) is

```c
size = r->path->length + (encode * 2) + 1 + r->args->length;
```

e.g

```
5 + (7*2) + 1 + 8 = 28
```

Comment (hongzhidao) [reply src/nxt_h1proto.c:2398]: Yes, I did.

Comment (ac000) [reply src/nxt_h1proto.c:2398]: Final question about this.

Why the * 2?

Comment (hongzhidao) [reply src/nxt_h1proto.c:2398]: According to its implementation.
https://github.com/nginx/unit/blob/master/src/nxt_string.c#L607
Take `/a b c` as an example, the allocation length is `r->path->length` + `encoded_number * 2`.
```
/a b c => /a%20b%20c
r->path: "/a b c"
encoded chars: ' ', ' '
total: 6 + 2 * 2
```

Comment (ac000) [reply src/nxt_h1proto.c:2398]: Got it.

r->path->length already includes the to be encoded chars, and each to be encoded char adds two bytes.

**Review by ac000 — APPROVED** [review]

File: src/nxt_h1proto.c (hunk 5 above)
Comment (hongzhidao) [reply src/nxt_h1proto.c:2398]: Correct.
